compiler/src/test/scala/org/scalajs/core/compiler/test/JSGlobalScopeTest.scala

```diff
@@ -0,0 +1,255 @@
+
+import org.scalajs.core.compiler.test.util._
+
+import org.junit.Test
+import org.junit.Ignore
+
+// scalastyle:off line.size.limit
+
+class JSGlobalScopeTest extends DirectTest with TestHelpers {
+
+  override def preamble: String = {
+    """
+    import scala.scalajs.js
+    import scala.scalajs.js.annotation._
+
+    @js.native
+    @JSGlobalScope
+    object SomeGlobalScope extends js.Any {
+      var validVar: Int = js.native
+      def validDef(): Int = js.native
+
+      var `not-a-valid-identifier-var`: Int = js.native
+      def `not-a-valid-identifier-def`(): Int = js.native
+
+      def +(that: Int): Int = js.native
+
+      @JSBracketAccess
+      def bracketSelect(name: String): Int = js.native
+      @JSBracketAccess
+      def bracketUpdate(name: String, v: Int): Unit = js.native
+
+      @JSBracketCall
+      def bracketCall(name: String)(arg: Int): Int = js.native
+
+      var arguments: js.Array[Any] = js.native
+      @JSName("arguments") def arguments2(x: Int): Int = js.native
+    }
+    """
+  }
+
+  @Test
+  def canAccessLegitMembers: Unit = {
+    s"""
+    object Main {
+      def main(): Unit = {
+        val a = js.Dynamic.global.validVar
+        js.Dynamic.global.validVar = 3
+        val b = js.Dynamic.global.validDef()
+
+        val c = SomeGlobalScope.validVar
+        SomeGlobalScope.validVar = 3
+        val d = SomeGlobalScope.validDef()
+
+        val e = SomeGlobalScope.bracketSelect("validVar")
+        SomeGlobalScope.bracketUpdate("validVar", 3)
+        val f = SomeGlobalScope.bracketCall("validDef")(4)
+      }
+    }
+    """.hasNoWarns
+  }
+
+  @Test
+  def noLoadGlobalValue: Unit = {
+    s"""
+    object Main {
+      def main(): Unit = {
+        val g1 = js.Dynamic.global
+        val g2 = SomeGlobalScope
+      }
+    }
+    """ hasErrors
+    s"""
+      |newSource1.scala:30: error: Loading the global scope as a value (anywhere but as the left-hand-side of a `.`-selection) is not allowed.
+      |  See https://www.scala-js.org/doc/interoperability/global-scope.html for further information.
+      |        val g1 = js.Dynamic.global
+      |                            ^
+      |newSource1.scala:31: error: Loading the global scope as a value (anywhere but as the left-hand-side of a `.`-selection) is not allowed.
+      |  See https://www.scala-js.org/doc/interoperability/global-scope.html for further information.
+      |        val g2 = SomeGlobalScope
+      |                 ^
+    """
+  }
+
+  @Test
+  def rejectInvalidJSIdentifiers: Unit = {
+    s"""
+    object Main {
+      def main(): Unit = {
+        val a = js.Dynamic.global.`not-a-valid-identifier-var`
+        js.Dynamic.global.`not-a-valid-identifier-var` = 3
+        val b = js.Dynamic.global.`not-a-valid-identifier-def`()
+
+        val c = SomeGlobalScope.`not-a-valid-identifier-var`
+        SomeGlobalScope.`not-a-valid-identifier-var` = 3
+        val d = SomeGlobalScope.`not-a-valid-identifier-def`()
```

Comment:
> I almost feel like we should disallow subclassing like this.

That would be extremely inconvenient for something like the DOM global scope, which really *inherits* from `Window`. You would have to duplicate the definition of all members in `Window` and in an `@JSGlobalScope` object.

The global scope does *inherit* from a class, and that includes `js.Object`. However, we cannot load it as a *value*. These things are not contradictory. Since we prevent from loading it as a value, in practice it indeed does not have "any type", since the *subtyping* relationship is always overruled. However, *inheritance* should be allowed.

> let's add definition site errors for direct definitions

I'm not sure. It means strictly more code in `PrepJSInterop` (there is no existing similar code that we can generalize for this) + it would be incomplete anyway (because of inheritance) + it is caught at use site anyway. Doesn't seem like something worth paying the maintenance cost for.